oximeter/db/src/oxql/ast/grammar.rs

```diff
@@ -0,0 +1,1125 @@
+        }
+
+        /// Parse a reducing operation by name.
+        pub rule reducer() -> Reducer
+            = "mean" { Reducer::Mean }
+            / "sum" { Reducer::Sum }
+            / expected!("a reducer name")
+
+        rule ws_with_comma() = _? "," _?
+        pub rule group_by() -> GroupBy
+            = "group_by"
+                _
+                "[" _? identifiers:(ident() ** ws_with_comma()) ","? _? "]"
+                reducer:("," _? red:reducer() { red })?
+        {
+            GroupBy {
+                identifiers,
+                reducer: reducer.unwrap_or_default(),
+            }
+        }
+
+        /// Parse a `join` table operation.
+        pub rule join() = "join" {}
+
+        pub(super) rule alignment_method() -> AlignmentMethod
+            = "interpolate" { AlignmentMethod::Interpolate }
+            / "mean_within" { AlignmentMethod::MeanWithin }
+
+        /// Parse an alignment table operation.
+        pub rule align() -> Align
+            = "align" _ method:alignment_method() "(" period:duration_literal_impl() ")"
+        {
+            Align { method, period }
+        }
+
+        pub(in super) rule basic_table_op() -> TableOp
+            = g:"get" _ t:timeseries_name() { TableOp::Basic(BasicTableOp::Get(t)) }
+            / f:filter() { TableOp::Basic(BasicTableOp::Filter(f)) }
+            / g:group_by() { TableOp::Basic(BasicTableOp::GroupBy(g)) }
+            / join() { TableOp::Basic(BasicTableOp::Join(Join)) }
+            / a:align() { TableOp::Basic(BasicTableOp::Align(a)) }
+
+        pub(in super) rule grouped_table_op() -> TableOp
+            = "{" _? ops:(query() ++ grouped_table_op_delim()) _? "}"
+        {
+            TableOp::Grouped(GroupedTableOp { ops })
+        }
+
+        /// Parse a top-level OxQL query.
+        ///
+        /// Queries always start with a "get" operation, and may be followed by
+        /// any number of other timeseries transformations
+        pub rule query() -> Query
+            = ops:(basic_table_op() / grouped_table_op()) ++ query_delim()
+        {?
+            let query = Query { ops };
+            if query.all_gets_at_query_start() {
+                Ok(query)
+            } else {
+                Err("every subquery must start with a `get` operation")
+            }
+        }
+
+        rule grouped_table_op_delim() = quiet!{ _? ";" _? }
+        rule query_delim() = quiet!{ _? "|" _? }
+    }
+}
+
+// Recognize escape sequences and convert them into the intended Unicode point
+// they represent.
+//
+// For example, the string containing ASCII "abcd" is returned unchanged.
+//
+// The string containing "\u{1234}" is returned as the string "ሴ". Note that the
+// Unicode bytes must be enclosed in {}, and can have length 1-6.
+//
+// If the string contains an invalid escape sequence, such as "\uFFFF", or a
+// control code, such as `\u07`, `None` is returned.
+fn recognize_escape_sequences(s: &str) -> Option<String> {
```

Comment:
If I'm understanding your question, it's more of the latter.

I'll admit this got a bit confusing to me at various points, so I could be doing unnecessary work. But here's my thought process. Suppose one want to write a query that matched a field in some timeseries, of type string, where that string contained a newline, `\n`. Say, `"here's a \n newline"`.

We get that from the public API, which means it's JSON, and so the newline is escaped. Since don't want to recognize the byte sequence `0x5c, 0x5c, 0x5c, 0x20` (`\\\n`), we need to _unescape_ it. The function `recognize_escape_sequences()` does that for the most common sequences I expected: newlines, tabs, form-feeds, nulls, literal backslashes, and Unicode points written in a Rust-syntax, like `\u{1234}`. The test `test_recognize_escape_sequences_with_valid_escape_sequence()` might help clarify. For example, the raw string `r#"\n"#` should compare equal to the Rust `char` '\n', i.e., an actual newline (`0x20`).

By the way, each time I look at this code, I convince myself I've implemented an expensive no-op, and then come back around to understanding why it's there! I really appreciate you taking the time to look at it, and please let me know if I _am_ actually confused :)

Reply:
I see. So this is almost like a pre-processor step to clean things up to make the actual grammar simpler and not have to know about things like escaped escape sequences haha. That makes total sense. Could be helpful to readers of the code to add something to that effect in the comment.

Reply:
Sounds good, will add a comment.

Reply:
Done in 08ede8a6a